rtio: MPSC queues for iodevs

teburd opened the pull request:
Rather than a fixed size queue, and with it the possibility of running
out of pre-allocated spots, each iodev now holds a wait-free mpsc
queue head.
    
This changes the parameter of iodev submit to be a struct
containing the 3 pointers for the rtio context, the submission queue entry,
and the queue pointer needed for the mpsc queue.
    
This solves the problem involving busy iodevs working with real
devices. For example a busy SPI bus driver could enqueue, without locking,
a request to start once the current request is done.
    
The queue entries are expected to be owned and allocated by the
executor rather than the iodev. This helps simplify potential
tuning knobs to one place, the RTIO context and its executor an
application directly uses.
    
As the test case shows iodevs can operate effectively lock free
with the mpsc queue and a single atomic denoting the current task.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Add lock free MPSC queue for iodevs and more

Adds a lock free/wait free MPSC queue to the rtio subsystem.

While the SPSC ring queue is fast and cache friendly it doesn't
work for all scenarios. Particularly the case where multiple rtio contexts
are attempting to work with a single iodev. An MPSC queue works perfectly
in this scenario.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Shareable lock-free iodevs

By using an mpsc queue for each iodev, the iodev itself is shareable across
contexts. Since its lock free, submits may occur even from an ISR context.

Rather than a fixed size queue, and with it the possibility of running
out of pre-allocated spots, each iodev now holds a wait-free mpsc
queue head.

This changes the parameter of iodev submit to be a struct containing 4
pointers for the rtio context, the submission queue entry, and the mpsc
node for the iodevs submission queue.

This solves the problem involving busy iodevs working with real
devices. For example a busy SPI bus driver could enqueue, without locking,
a request to start once the current request is done.

The queue entries are expected to be owned and allocated by the
executor rather than the iodev. This helps simplify potential
tuning knobs to one place, the RTIO context and its executor an
application directly uses.

As the test case shows iodevs can operate effectively lock free
with the mpsc queue and a single atomic denoting the current task.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

nordic-krch reviewed (commented):
It looks ok.

teburd commented on tests/subsys/rtio/rtio_api/src/main.c:
```diff
@@ -231,6 +234,171 @@ ZTEST(rtio_spsc, test_spsc_threaded)
+}
+
+/**
+ * @brief Test that the producer and consumer are indeed thread safe
```
That's a great idea, I'll give it a whirl. Thanks for taking a look!

nordic-krch commented:
Few thoughts regarding rtio future features:
- For i2c we probably need an option to combine multiple events into one. There is TX-RX or TX-TX with repeated start, etc. Currently there is a single mpsc sqe per rtio so it's not possible to schedule more than one. Maybe some chaining of events could be added (mpsc_sqe pointer to rtio_sqe which points to next rtio_sqe, etc.)

- I would like to be able to get entry in `cq` on the last event. Let say i do an initialization of a sensor which contains 10 i2c writings. I don't want to wake up the thread on every completion (by the semaphore) but only on the last one or error. It could probably be achieved through some flags.

- It would be nice to get some instrumentation for measuring spsc utilization.

- I would like to be able to add a delay as one of the events to be able to execute something like trigger sensor, wait for some time, read the results (wake up when full sequence is completed see 2. ). This could be done through some iodev using k_timer or scheduled k_work that would sort delay requests.

- How i2c dev id will be passed to the i2c iodev?

- How iodev specific flags are passed?

teburd commented:
> Few thoughts regarding rtio future features:
> 
>     * For i2c we probably need an option to combine multiple events into one. There is TX-RX or TX-TX with repeated start, etc. Currently there is a single mpsc sqe per rtio so it's not possible to schedule more than one. Maybe some chaining of events could be added (mpsc_sqe pointer to rtio_sqe which points to next rtio_sqe, etc.)

I agree, there's a few options here but let me explain what I think is the simplest

Submissions may be chained today, already, with a flag. Mostly this is for error handling/ordering of operations across devices though rather than creating a transaction. This is where its a bit tricky if you want to use that chaining as a transaction on one device.

It's probably simpler to avoid trying to use this chaining for a transaction. Instead, like io_uring itself does, add new op codes to represent those unique equivalent vector i/o like syscalls for i2c_transfer/spi_transceive.

This just makes things so much simpler if done this way rather than trying to re-orient the existing APIs as a queue of individual operations.

As an interesting side effect if new RTIO_OP_I2C_TRANSFER/RTIO_OP_SPI_TRANSCEIVE ops are added, each call to i2c_transfer/spi_transceive APIs could allocate a tiny rtio context and add its operation to the queue, waiting on a semaphore to be given. The async version would be a little more interesting of course because something would need to be allocated.

This is an experiment though and truly ideas and feedback would be really helpful.

> 
>     * I would like to be able to get entry in `cq` on the last event. Let say i do an initialization of a sensor which contains 10 i2c writings. I don't want to wake up the thread on every completion (by the semaphore) but only on the last one or error. It could probably be achieved through some flags.

You mentioned before you'd like to get a callback as well. The answer would be add RTIO op codes and chain your operations into the one you want to wait on. An RTIO_OP_SEM_GIVE could be available where the struct rtio_iodev is replaced with a struct k_sem *, the executors would need to understand how to do this more directly which isn't a bad thing. There's other OPs (delays, deadlines, callbacks) that would be useful in the same way. Some common code could be available.

The way the current rtio_submit() and rtio_cqe_consume_block() calls work today with a semaphore *is* less than ideal. But the API doesn't stipulate that it uses a semaphore internally, it just happened to be a quick and easy way of making it work.

Ideally struct rtio would have its own _wait_q and semantics around it for doing what rtio_submit does today. Alternatively you could do the above, not wait at all on the rtio context and add a RTIO_OP_SEM_GIVE with a semaphore. Some better logic on what to do when the previous SQE errors would need to happen to make that work the way we'd all I think expect it to.

> 
>     * It would be nice to get some instrumentation for measuring spsc utilization.

What sort of instrumentation are you looking for, similar to k_thread stack usage with current/max kind of thing?

> 
>     * I would like to be able to add a delay as one of the events to be able to execute something like trigger sensor, wait for some time, read the results (wake up when full sequence is completed see 2. ). This could be done through some iodev using k_timer or scheduled k_work that would sort delay requests.
> 

Absolutely part of the known missing features. RTIO_OP_DELAY with a k_timer object rather than a struct rtio_iodev would be needed for that. The same is true of work queues. RTIO_OP_WORK_SUBMIT could exist, and happen at the tail end or start of a chain or both.

>     * How i2c dev id will be passed to the i2c iodev?

My basic plan was something like...

``` c
struct my_driver_data {
#ifdef CONFIG_RTIO
  struct rtio_iodev iodev;
#endif
};

int my_driver_iodev_submit(...) 
{
   struct my_driver_data *drv_data = CONTAINER_OF(..);

}
```

> 
>     * How iodev specific flags are passed?

A member could be added to rtio_sqe to contain iodev_flags if needed. The size of this struct, while using RAM isn't likely to be a huge deal in my opinion as its likely each context is going to have a small queue of them, with perhaps a few contexts per application for small devices.

teburd commented:
Some neat ideas came from a small discussion group.

Re-using the userdata pointer to point to the next sqe in the chain was one, and was really interesting. By doing that the device queue can have transactions atomically added in one single mpsc queue insert.

Some further thoughts to this... there is one potential issue in that not all rtio_sqe's in the chain are guaranteed to operate on the same device at the moment. That could be a restriction that is enforced though if it greatly simplifies the common uses, which I think it would.

teburd commented:
I'm pushing off changes to enable polling and the transactional work mentioned. This gets a big step forward

I have a in progress branch with a real board and real sensor where I'm working towards enabling RTIO with SPI... this is the stable set of changes that I need for that.

See https://github.com/teburd/zephyr/tree/tdk_icm42688p_rtio for how this is used with a real sensor at the moment (thought with a thread to do SPI still since I haven't finished that bit just yet)

MaureenHelm commented on samples/subsys/rtio/sensor_batch_processing/src/vnd_sensor.c:
```diff
@@ -65,36 +65,28 @@ static void vnd_sensor_iodev_execute(const struct device *dev,
-		LOG_ERR("%s: Could not put a msg", dev->name);
-		rtio_sqe_err(r, sqe, -EWOULDBLOCK);
-	}
+	struct vnd_sensor_data *data = (struct vnd_sensor_data *) iodev_sqe->sqe->iodev;
+
+	rtio_mpsc_push(&data->iodev.iodev_sq, &iodev_sqe->q);
```
This is nice

Commit:
rtio: Remove unused spsc initializer params

SPSC initializer macro had unused parameters defined and documented. Remove
those as they are unused.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented on include/zephyr/rtio/rtio_spsc.h:
```diff
@@ -75,11 +75,9 @@ struct rtio_spsc {
  * @param sz Size of the spsc, must be power of 2 (ex: 2, 4, 8)
  */
-#define RTIO_SPSC_INITIALIZER(name, type, sz)	  \
+#define RTIO_SPSC_INITIALIZER(sz)		  \
```
split into another commit

Commit:
rtio: Exclude failing renode platform

Renode platform fails the test despite it working well on qemu riscv.
Ignore this particular platform for now.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
rtio: Drop stray printks in API test

Noticed the tests were a bit verbose, saw a few stray printks. Drop those
as they aren't really needed and potentially cause testing issues, printk
is a potential synchronization point.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
I ended up excluding a renode platform that fails the test for now. Something to look at later at some point. Also cleaned up and removed a few printks in the tests.

teburd commented:
I have a set of changes waiting on this that adds transactions (for i2c/spi) and ongoing work adding SPI support and an initial driver with the sam spi driver on the tdk robokit1.

@yperess @nordic-krch please re-review when you have a chance, thanks!

teburd commented on include/zephyr/rtio/rtio_mpsc.h:
```diff
@@ -0,0 +1,156 @@
+/*
```
Yeah definitely, the two atomic queues are useful beyond rtio.

teburd commented on include/zephyr/rtio/rtio_executor_simple.h:
```diff
@@ -36,26 +36,25 @@ int rtio_simple_submit(struct rtio *r);
+	struct rtio_iodev_sqe task;
 };
 
 /**
```
This is a good point, thanks! Will create a PR that fixes this.
